### ISSUE TITLE
Updating & Commitment Loss

### ISSUE THREAD

**Issue by goldbird5** [lucidrains/vector-quantize-pytorch#49]
Hi, I have some questions about recent commits.

1) Recent [commit](https://github.com/lucidrains/vector-quantize-pytorch/commit/d2ec3998e021821ea9951a3d0d13b23ccbca2e95) modified to get quantized vector after updating 'self.embed'. Then, I wonder whether we should get new distances after updating.
I mean, we use 'embed_ind' to get quantize, and calculate 'dist' to get 'embed_ind'. However, updated codebook might give different 'dist' and 'embed_ind'. How do you think about calculating dist again after updating codebook and before getting quantize vectors?

2) Also, this [commit](https://github.com/lucidrains/vector-quantize-pytorch/commit/8417aeced5e1be4a231f40e03046921be2955298) uses 'l2norm_x' to get quantize using cosine sim. I think this was intended to make original input 'x' remains in its original domain, and gradient loss would be backpropagated through 'l2norm_x'. But still, commitment loss is calculated using original x and l2-normed quantize, which push input x to l2-normed hypersphere. Should we calculate commitment loss also with 'l2norm_x'?

Thanks in advance.

**Comment by amitaie** [comment]
I think @goldbird5 right in his first point, we have to calc the embed_ind again because otherwise we will change the self.embed but not the embed_ind so in the "batched_embedding" we will take vectors that is not related at all to the input one. 

But i'm not sure this commit is right because if we think of this RVQ as some NN, in the forward we quantize with the weights  that we have so far, and proceed with the results of it to the end, and we change our weights only for the next batch.

**Comment by goldbird5** [comment]
Actually, I personally agreed with the statement of commit, 'quantized should be selected after EMA', since updated codebook might be containing better codes, and as mentioned in message of commit, commitment loss could push input x to updated codes.

In case of general NN, we should inevitably update weights after foward process of current batch because we should use gradients backpropagated from final loss. But in case of that we can update some variables during forward process instantly, (e.g. EMA update of codebook in here) I think updating and utilizing updated codes instantly could be reasonable.

**Comment by lucidrains** [comment]
@goldbird5 just ran an experiment with the new cosine sim with smoothing of cluster size and sum of embeddings and it seems to work ok!

**Comment by goldbird5** [comment]
@lucidrains Sounds Good! I also tried using updated codes to get quantize(version before reverting), and it looks also ok.

**Comment by goldbird5** [comment]
@lucidrains By the way, I have some question about recent straight-through implementaion.
Since we use 'one_hot' only for updating codebook with EMA, and output of entire model is

quantize = batched_embedding(embed_ind, self.embed)
...
quantize = x + (quantize-x).detach(),

there seems no effect of straight-through for one-hot. Gradient might be only backpropagated through input x.

Is there something I'm missing?

**Comment by lucidrains** [comment]
@goldbird5 :man_facepalming: i was just testing to see if you were paying attention (one of my high school teacher's favorite line) :laughing: 

let me go fix it now and rerun some experiments. even more embarrassing, it means some of the experiments i was interpreting yesterday was just normal variance between runs.

**Comment by lucidrains** [comment]
closing because i think it was all addressed
